File: binary-trees/flip-binary-tree-to-match-preorder-traversal/flip-binary-tree-to-match-preorder-traversal.go
package flipbinarytreetomatchpreordertraversal

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var i int

func flipMatchVoyage(root *TreeNode, voyage []int) []int {
	i = 0
	list := new([]int)

	if dfs(root, voyage, list) {
		return *list
	}
	return []int{-1}
}

func dfs(node *TreeNode, voy []int, list *[]int) bool {
	if node == nil {
		return true
	}

	if node.Val != voy[i] {
		return false
	}

	i++

	if node.Left != nil && node.Left.Val != voy[i] {
		*list = append(*list, node.Val)
		return dfs(node.Right, voy, list) && dfs(node.Left, voy, list)
	}

	return dfs(node.Left, voy, list) && dfs(node.Right, voy, list)
}

File: binary-trees/flip-binary-tree-to-match-preorder-traversal/flip-binary-tree-to-match-preorder-traversal.java
class Solution {
  int i = 0;

  public List<Integer> flipMatchVoyage(TreeNode root, int[] voyage) {
    List<Integer> list = new ArrayList<>();
    if (dfs(root, voyage, list)) {
      return list;
    }

    return Arrays.asList(-1);
  }

  private boolean dfs(TreeNode node, int[] voy, List<Integer> list) {
    if (node == null) {
      return true;
    }

    if (node.val != voy[i]) {
      return false;
    }

    i++;

    if (node.left != null && node.left.val != voy[i]) {
      list.add(node.val);
      return dfs(node.right, voy, list) && dfs(node.left, voy, list);
    }

    return dfs(node.left, voy, list) && dfs(node.right, voy, list);
  }
}
